File: backend/app/core/logger.py
import logging
import os
from logging import LoggerAdapter
from app.core.config import get_settings  # Load project configuration

# Fetch settings from .env
settings = get_settings()
LOG_DIR = "logs"
os.makedirs(LOG_DIR, exist_ok=True)

# Primary logger
base_logger = logging.getLogger("app")

# Log levels controlled via ENV
LOG_LEVELS = {
    "debug": logging.DEBUG,
    "info": logging.INFO,
    "warning": logging.WARNING,
    "error": logging.ERROR,
    "critical": logging.CRITICAL
}
log_level = LOG_LEVELS.get(settings.debug_level.lower(), logging.INFO)
base_logger.setLevel(log_level)

# Custom formatter that aligns level labels and source tags
class AlignedFormatter(logging.Formatter):
    LEVEL_FORMATS = {
        logging.DEBUG:    "DEBUG",
        logging.INFO:     "INFO ",
        logging.WARNING:  "WARN ",
        logging.ERROR:    "ERROR",
        logging.CRITICAL: "CRIT "
    }

    def format(self, record):
        record.levelname = self.LEVEL_FORMATS.get(record.levelno, record.levelname)
        return super().format(record)

class SourceAwareFormatter(AlignedFormatter):
    def format(self, record):
        source = getattr(record, "source", "")
        if source:
            src = str(source).upper()
            # Limit to four characters; pad if shorter
            src = src[:4].ljust(4)
            source_tag = f"[{src}]"
            if not str(record.msg).startswith(source_tag):
                record.msg = f"{source_tag} {record.msg}"
        return super().format(record)

formatter = SourceAwareFormatter("%(asctime)s - %(levelname)s - %(message)s")

# Console handler
console_handler = logging.StreamHandler()
console_handler.setLevel(logging.INFO if settings.app_env == "production" else log_level)
console_handler.setFormatter(formatter)

base_logger.addHandler(console_handler)

# Disable propagation up to root (FastAPI/Uvicorn already manage their own)
base_logger.propagate = False

# Helper to get a logger with a source context
def get_logger(source: str = "") -> LoggerAdapter:
    return LoggerAdapter(base_logger, {"source": source})

# Default global logger
logger = get_logger()

# Sample startup log
logger.info(f"Logger initialized with level: {settings.debug_level} (env: {settings.app_env})")

File: frontend/src/utils/logger.ts
// src/utils/logger.ts

export enum LogLevel {
  DEBUG = 10,
  INFO = 20,
  WARN = 30,
  ERROR = 40,
  NONE = 100,
}

const LEVEL_NAMES: Record<LogLevel, string> = {
  [LogLevel.DEBUG]: "DEBUG",
  [LogLevel.INFO]:  "INFO ",
  [LogLevel.WARN]:  "WARN ",
  [LogLevel.ERROR]: "ERROR",
  [LogLevel.NONE]:  "NONE ",
}

// читаем уровень из окружения (vite: VITE_LOG_LEVEL)
const ENV_LOG_LEVEL = (import.meta.env.VITE_LOG_LEVEL || "info").toLowerCase()
const LOG_LEVELS: Record<string, LogLevel> = {
  debug: LogLevel.DEBUG,
  info: LogLevel.INFO,
  warn: LogLevel.WARN,
  error: LogLevel.ERROR,
  none: LogLevel.NONE,
}
let currentLevel = LOG_LEVELS[ENV_LOG_LEVEL] ?? LogLevel.INFO

function shouldLog(level: LogLevel): boolean {
  return level >= currentLevel
}

function padSource(source: string): string {
  // всегда 4 символа: WS, APP, API → [WS  ], [APP], [API]
  if (source.length > 4) {
    return source.slice(0, 4).toUpperCase()
  }
  return source.toUpperCase().padEnd(4, " ")
}

function toStr(msg: any): string {
  if (typeof msg === "string") return msg
  try {
    return JSON.stringify(msg)
  } catch {
    return String(msg)
  }
}

function log(level: LogLevel, source: string, message: any, ...optionalParams: any[]) {
  if (!shouldLog(level)) return

  const levelName = LEVEL_NAMES[level]
  const sourceTag = `[${padSource(source)}]`
  const formatted = `${levelName} ${sourceTag} ${toStr(message)}`

  switch (level) {
    case LogLevel.DEBUG: console.debug(formatted, ...optionalParams); break
    case LogLevel.INFO:  console.info(formatted, ...optionalParams); break
    case LogLevel.WARN:  console.warn(formatted, ...optionalParams); break
    case LogLevel.ERROR: console.error(formatted, ...optionalParams); break
  }
}

export function getLogger(source: string) {
  return {
    debug: (msg: any, ...args: any[]) => log(LogLevel.DEBUG, source, msg, ...args),
    info:  (msg: any, ...args: any[]) => log(LogLevel.INFO,  source, msg, ...args),
    warn:  (msg: any, ...args: any[]) => log(LogLevel.WARN,  source, msg, ...args),
    error: (msg: any, ...args: any[]) => log(LogLevel.ERROR, source, msg, ...args),
  }
}

export const logger = getLogger("APP")

logger.info(
  `🗒️ Logger initialized with level=${ENV_LOG_LEVEL} (mode=${import.meta.env.MODE})`
)
